gossip/server.go

```diff
@@ -99,45 +101,49 @@ func (s *server) Gossip(argsI proto.Message) (proto.Message, error) {
 	// as a permanent peer. We always accept its input and return
 	// our delta.
 	canAccept := true
-	if !s.incoming.hasNode(args.NodeID) {
-		s.incoming.setMaxSize(s.maxPeers())
-		if !s.incoming.hasSpace() {
-			canAccept = false
+	if args.NodeID != 0 {
```

Comment:
The changes in this file are actually really minimal–it's basically rearranging things so we no longer trade gossip if the connection will be forwarded (doing so made computation of hops between any two nodes imaccurate).